File: src/core/vits_driver.cpp
#include "loom/core/vits_driver.h"

#include "loom/core/relative_position.h"
#include "loom/loom_errors.h"

#include <ggml-backend.h>

#include <algorithm>
#include <cmath>
#include <random>

namespace loom {

VitsDriver::VitsDriver(GgufModel& stats_model, GraphTopology stats_topo, GgufModel& logw_model,
                        GraphTopology logw_topo, GgufModel& flow_vocoder_model, GraphTopology flow_vocoder_topo,
                        VitsConfig cfg, ggml_backend_t backend)
    : stats_model_(stats_model),
      logw_model_(logw_model),
      flow_vocoder_model_(flow_vocoder_model),
      cfg_(cfg),
      backend_(backend),
      stats_topo_(std::move(stats_topo)),
      logw_topo_(std::move(logw_topo)),
      flow_vocoder_topo_(std::move(flow_vocoder_topo)) {
    stats_builder_ = std::make_unique<GraphBuilder>(stats_topo_, stats_model_, backend_, /*kv_cache=*/nullptr);
    logw_builder_ = std::make_unique<GraphBuilder>(logw_topo_, logw_model_, backend_, /*kv_cache=*/nullptr);
    flow_vocoder_builder_ =
        std::make_unique<GraphBuilder>(flow_vocoder_topo_, flow_vocoder_model_, backend_, /*kv_cache=*/nullptr);
}

std::vector<float> VitsDriver::synthesize(const std::vector<int32_t>& token_ids, uint32_t seed) {
    const auto T = static_cast<uint32_t>(token_ids.size());
    if (T == 0) {
        throw Error("VitsDriver::synthesize: token_ids must be non-empty");
    }
    const int64_t k_channels = cfg_.hidden_channels / cfg_.n_heads;
    const int64_t inter_channels = cfg_.inter_channels;
    std::mt19937 rng(seed);
    std::normal_distribution<float> normal(0.0f, 1.0f);

    // Shared per-call fill logic: tokens, an all-zero (no padding) additive attention mask, and every
    // TextEncoder layer's dynamic-T relative-position tables -- identical for both the `stats` and
    // `logw` builds (both recompute TextEncoder from scratch; see BACKLOG.md's "why two separate
    // topologies" note).
    auto fill_text_encoder_inputs = [&](GgufModel& model, GraphBuilder::BuildResult& r) {
        std::vector<int32_t> tokens_copy = token_ids;
        ggml_backend_tensor_set(r.input_tensors.at("tokens"), tokens_copy.data(), 0, tokens_copy.size() * sizeof(int32_t));
        std::vector<float> mask(static_cast<size_t>(T) * T, 0.0f);
        ggml_backend_tensor_set(r.input_tensors.at("attn_mask"), mask.data(), 0, mask.size() * sizeof(float));

        for (uint32_t i = 0; i < cfg_.n_text_layers; ++i) {
            const std::string prefix = "enc_p.encoder.attn_layers." + std::to_string(i);
            for (const char* which : {"emb_rel_k", "emb_rel_v"}) {
                ggml_tensor* raw_t = model.weight(prefix + "." + which + "_raw");
                std::vector<float> raw(static_cast<size_t>(ggml_nelements(raw_t)));
                ggml_backend_tensor_get(raw_t, raw.data(), 0, raw.size() * sizeof(float));
                std::vector<float> table = pad_crop_relative_embeddings(raw, cfg_.window_size, k_channels, T);
                ggml_backend_tensor_set(r.input_tensors.at(std::string(which) + "_" + std::to_string(i)), table.data(),
                                         0, table.size() * sizeof(float));
            }
        }
    };

    // --- Phase 1a: stats = TextEncoder -> [m_p; logs_p] (channel-first, [2*inter_channels, T]) ---
    GraphBuilder::BuildResult stats_r = stats_builder_->build({{"n_tokens", T}, {"n_past", 0}});
    fill_text_encoder_inputs(stats_model_, stats_r);
    ggml_backend_graph_compute(backend_, stats_r.graph);
    std::vector<float> stats(static_cast<size_t>(ggml_nelements(stats_r.output)));
    ggml_backend_tensor_get(stats_r.output, stats.data(), 0, stats.size() * sizeof(float));

    // --- Phase 1b: logw = TextEncoder + StochasticDurationPredictor(reverse) -> [T] duration logits ---
    GraphBuilder::BuildResult logw_r = logw_builder_->build({{"n_tokens", T}, {"n_past", 0}});
    fill_text_encoder_inputs(logw_model_, logw_r);
    std::vector<float> z_noise(static_cast<size_t>(T) * 2);
    for (float& v : z_noise) v = normal(rng) * cfg_.noise_scale_w;
    ggml_backend_tensor_set(logw_r.input_tensors.at("z_noise"), z_noise.data(), 0, z_noise.size() * sizeof(float));
    ggml_backend_graph_compute(backend_, logw_r.graph);
    std::vector<float> logw(T);
    ggml_backend_tensor_get(logw_r.output, logw.data(), 0, logw.size() * sizeof(float));

    // --- Host-side generate_path: w_ceil[t] = ceil(exp(logw[t]) * length_scale); y_length =
    //     max(sum(w_ceil), 1) -- real models.py::infer(), with x_mask/y_mask dropped (both always
    //     all-ones: single unpadded utterance, no batching, matching every other simplification this
    //     whole VITS effort has made). The alignment matrix itself degenerates to a plain "replicate
    //     column t of m_p/logs_p for w_ceil[t] consecutive output frames" expansion once the mask is
    //     gone -- computed directly below, no explicit attn matrix ever materialized. ---
    std::vector<uint32_t> w_ceil(T);
    uint64_t y_length_u64 = 0;
    for (uint32_t t = 0; t < T; ++t) {
        const float w = std::exp(logw[t]) * cfg_.length_scale;
        w_ceil[t] = static_cast<uint32_t>(std::ceil(w));
        y_length_u64 += w_ceil[t];
    }
    const uint32_t y_length = static_cast<uint32_t>(std::max<uint64_t>(y_length_u64, 1));

    // z_p: [y_length, inter_channels], T-major (matches the coupling-flow/vocoder's own [T,C]
    // convention established throughout this VITS effort). m_p[c,t] = stats[t*(2*inter_channels)+c],
    // logs_p[c,t] = stats[t*(2*inter_channels)+inter_channels+c] (stats is channel-first, [C,T]).
    std::vector<float> z_p(static_cast<size_t>(y_length) * inter_channels);
    uint32_t out_frame = 0;
    for (uint32_t t = 0; t < T && out_frame < y_length; ++t) {
        const float* m_p_col = stats.data() + static_cast<size_t>(t) * 2 * inter_channels;
        const float* logs_p_col = m_p_col + inter_channels;
        for (uint32_t rep = 0; rep < w_ceil[t] && out_frame < y_length; ++rep, ++out_frame) {
            float* dst = z_p.data() + static_cast<size_t>(out_frame) * inter_channels;
            for (int64_t c = 0; c < inter_channels; ++c) {
                dst[c] = m_p_col[c] + normal(rng) * std::exp(logs_p_col[c]) * cfg_.noise_scale;
            }
        }
    }

    // --- Phase 2: coupling flow (reverse) + HiFi-GAN vocoder -> waveform ---
    GraphBuilder::BuildResult wav_r = flow_vocoder_builder_->build({{"n_tokens", y_length}, {"n_past", 0}});
    ggml_backend_tensor_set(wav_r.input_tensors.at("z_p"), z_p.data(), 0, z_p.size() * sizeof(float));
    ggml_backend_graph_compute(backend_, wav_r.graph);
    std::vector<float> waveform(static_cast<size_t>(ggml_nelements(wav_r.output)));
    ggml_backend_tensor_get(wav_r.output, waveform.data(), 0, waveform.size() * sizeof(float));
    return waveform;
}

} // namespace loom

File: tools/convert_piper_vits/vits_driver.lua
-- Lua port of loom::VitsDriver::synthesize (src/core/vits_driver.cpp), validating the
-- procedural-generalization architecture (LOOM_PROCEDURAL_GENERALIZATION.md / LOOM_MIL_CONVERSION.md)
-- against the most involved of the three CFM/flow-style drivers ported so far: dynamic-length
-- relative-position table cropping (loom.pad_crop_relative_embeddings + loom.get_weight, since the raw
-- learned tables are read directly off the GGUF weight table, not through any topology's own
-- declared inputs/outputs) combined with Gaussian-noise-interleaved duration expansion.
--
-- Expects three modules pre-registered by the host, ALL sharing the SAME underlying GgufModel (one
-- combined vits.gguf): "stats", "logw", "flow_vocoder" -- none use a KvCache.
--
-- inputs: token_ids (int array, TextEncoder's own vocabulary), seed (int, seeds loom.gaussian_array --
-- SDP's own z_noise AND z_p's own noise, the only two stochastic points in this pipeline, matching the
-- real driver's own single `rng_`), plus the real model constants n_text_layers, window_size,
-- k_channels, inter_channels, noise_scale, noise_scale_w, length_scale (VitsConfig's own real
-- defaults).
--
-- Returns: the raw waveform (flat f32 array), same convention as VitsDriver::synthesize's own return.
function synthesize(inputs)
    loom.seed_rng(inputs.seed)

    local T = #inputs.token_ids
    local attn_mask = loom.zero_mask(T, T)

    -- Shared per-call fill logic: tokens, attn_mask, and every TextEncoder layer's dynamic-T
    -- relative-position tables -- identical for both the "stats" and "logw" calls (matches
    -- VitsDriver::synthesize's own fill_text_encoder_inputs lambda).
    local function text_encoder_inputs()
        local tbl = { tokens = inputs.token_ids, attn_mask = attn_mask }
        for i = 0, inputs.n_text_layers - 1 do
            local raw_k = loom.get_weight("stats", "enc_p.encoder.attn_layers." .. i .. ".emb_rel_k_raw")
            local raw_v = loom.get_weight("stats", "enc_p.encoder.attn_layers." .. i .. ".emb_rel_v_raw")
            tbl["emb_rel_k_" .. i] = loom.pad_crop_relative_embeddings(raw_k, inputs.window_size, inputs.k_channels, T)
            tbl["emb_rel_v_" .. i] = loom.pad_crop_relative_embeddings(raw_v, inputs.window_size, inputs.k_channels, T)
        end
        return tbl
    end

    -- --- Phase 1a: stats = TextEncoder -> [m_p;logs_p] (channel-first [2*inter_channels,T] -- flat
    --     layout is ALREADY "T rows of 2*inter_channels contiguous floats", matching the C++ driver's
    --     own stats[t*2*inter_channels+c] indexing directly, no transpose needed). ---
    local stats = loom.run_subgraph("stats", T, 0, text_encoder_inputs())

    -- --- Phase 1b: logw = TextEncoder + StochasticDurationPredictor(reverse) -> [T] duration logits ---
    local z_noise = loom.gaussian_array(T * 2)
    for i = 1, #z_noise do z_noise[i] = z_noise[i] * inputs.noise_scale_w end
    local logw_inputs = text_encoder_inputs()
    logw_inputs.z_noise = z_noise
    local logw = loom.run_subgraph("logw", T, 0, logw_inputs)

    -- --- Host generate_path: w_ceil[t]=ceil(exp(logw[t])*length_scale); y_length=max(sum(w_ceil),1).
    --     Degenerates to a plain "replicate column t of m_p/logs_p for w_ceil[t] frames" expansion once
    --     x_mask/y_mask are dropped (single unpadded utterance) -- computed directly below, same as the
    --     C++ driver, no explicit attn matrix ever materialized. ---
    local w_ceil = {}
    local y_length = 0
    for t = 1, T do
        w_ceil[t] = math.ceil(math.exp(logw[t]) * inputs.length_scale)
        y_length = y_length + w_ceil[t]
    end
    if y_length < 1 then y_length = 1 end

    -- --- z_p: [y_length,inter_channels] (T-major [T,C]) -- m_p[c,t]=stats[t*2C+c],
    --     logs_p[c,t]=stats[t*2C+C+c]. ---
    local inter_channels = inputs.inter_channels
    local z_p = {}
    local out_frame = 0
    local noise_idx = 0
    local gaussian_pool = loom.gaussian_array(y_length * inter_channels)
    for t = 0, T - 1 do
        if out_frame >= y_length then break end
        local m_p_base = t * 2 * inter_channels
        local logs_p_base = m_p_base + inter_channels
        for rep = 1, w_ceil[t + 1] do
            if out_frame >= y_length then break end
            for c = 0, inter_channels - 1 do
                noise_idx = noise_idx + 1
                z_p[out_frame * inter_channels + c + 1] =
                    stats[m_p_base + c + 1] + gaussian_pool[noise_idx] * math.exp(stats[logs_p_base + c + 1]) * inputs.noise_scale
            end
            out_frame = out_frame + 1
        end
    end

    -- --- Phase 2: coupling flow (reverse) + HiFi-GAN vocoder -> waveform ---
    local waveform = loom.run_subgraph("flow_vocoder", y_length, 0, { z_p = z_p })
    return waveform
end
